google/cloud/storage/internal/rest_client.cc

```diff
@@ -288,9 +297,240 @@ StatusOr<BucketMetadata> RestClient::LockBucketRetentionPolicy(
       std::move(builder).BuildRequest(), {absl::MakeConstSpan(std::string{})}));
 }
 
+std::string RestClient::PickBoundary(std::string const& text_to_avoid) {
+  // We need to find a string that is *not* found in `text_to_avoid`, we pick
+  // a string at random, and see if it is in `text_to_avoid`, if it is, we grow
+  // the string with random characters and start from where we last found a
+  // the candidate.  Eventually we will find something, though it might be
+  // larger than `text_to_avoid`.  And we only make (approximately) one pass
+  // over `text_to_avoid`.
+  auto generate_candidate = [this](int n) {
+    static auto const* const kChars = new std::string(
+        "abcdefghijklmnopqrstuvwxyz0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ");
+    std::unique_lock<std::mutex> lk(mu_);
+    return google::cloud::internal::Sample(generator_, n, *kChars);
+  };
+  constexpr int kCandidateInitialSize = 16;
+  constexpr int kCandidateGrowthSize = 4;
+  return GenerateMessageBoundary(text_to_avoid, std::move(generate_candidate),
+                                 kCandidateInitialSize, kCandidateGrowthSize);
+}
+
+StatusOr<ObjectMetadata> RestClient::InsertObjectMediaMultipart(
+    InsertObjectMediaRequest const& request) {
+  RestRequestBuilder builder(
+      absl::StrCat("upload/storage/", options_.get<TargetApiVersionOption>(),
+                   "/b/", request.bucket_name(), "/o"));
+  auto auth = AddAuthorizationHeader(options_, builder);
+  if (!auth.ok()) return auth;
+
+  AddOptionsWithSkip<RestRequestBuilder, ContentType> no_content_type{builder};
+  request.ForEachOption(no_content_type);
+
+  if (request.HasOption<UserIp>()) {
+    builder.AddQueryParameter(UserIp::name(),
+                              request.GetOption<UserIp>().value());
+  }
+
+  // 2. Pick a separator that does not conflict with the request contents.
+  auto boundary = PickBoundary(request.contents());
+  builder.AddHeader("content-type", "multipart/related; boundary=" + boundary);
+  builder.AddQueryParameter("uploadType", "multipart");
+  builder.AddQueryParameter("name", request.object_name());
+
+  // 3. Perform a streaming upload because computing the size upfront is more
```

Comment:
nit: s/Perform a streaming upload/Use a `std::ostringstream` to compute the full payload/